datarecording/datarecorder.go

```diff
@@ -0,0 +1,260 @@
+	Init()
+
+	//CreateTable creates a new table with given filename
+	CreateTable(table string, sampleEntry any)
+
+	//DataInsert writes a same-type task into table that already exists
+	DataInsert(table string, entry any)
+
+	//ListTable returns a slice containing names of all tables
+	ListTables() []string
+
+	//Flush flushes all the baffered task into database
+	Flush()
+}
+
+// SQLiteWriter is the writer that writes data into SQLite database
+type SQLiteWriter struct {
+	*sql.DB
+	statement *sql.Stmt
+
+	dbName     string
+	tables     map[string][]any
+	batchSize  int
+	tableCount int
+	entryCount int
+}
+
+// NewSQLiteWriter creates a new SQLiteWriter.
+func NewSQLiteWriter(path string) *SQLiteWriter {
+	w := &SQLiteWriter{
+		dbName:    path,
+		batchSize: 100000,
+		tables:    make(map[string][]any),
+	}
+
+	atexit.Register(func() { w.Flush() })
+
+	return w
+}
+
+// Init establishes a connection to the databse
+func (t *SQLiteWriter) Init() {
+	if t.dbName == "" {
+		t.dbName = "akita_data_recording_" + xid.New().String()
+	}
+
+	filename := t.dbName + ".sqlite3"
+	_, err := os.Stat(filename)
+	if err == nil {
+		panic(fmt.Errorf("file %s already exists", filename))
+	}
+
+	fmt.Fprintf(os.Stderr, "Database created for recording: %s\n", filename)
+
+	db, err := sql.Open("sqlite3", filename)
+	if err != nil {
+		panic(err)
+	}
+
+	t.DB = db
+}
+
+func (t *SQLiteWriter) isAllowedType(kind reflect.Kind) bool {
+	switch kind {
+	case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr,
+		reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128,
+		reflect.String, reflect.UnsafePointer:
+		return true
+	default:
+		return false
+	}
+}
+
+func (t *SQLiteWriter) checkStructFields(entry any) error {
+	types := reflect.TypeOf(entry)
+
+	for i := 0; i < types.NumField(); i++ {
+		field := types.Field(i)
+
+		fieldKind := field.Type.Kind()
+		if !t.isAllowedType(fieldKind) {
+			return errors.New("entry is invalid")
+		}
+	}
+	return nil
+}
+
+func (t *SQLiteWriter) CreateTable(table string, sampleEntry any) {
+	err := t.checkStructFields(sampleEntry)
+	if err != nil {
+		panic(err)
+	}
+
+	t.tableCount++
+	n := structs.Names(sampleEntry)
+	fields := strings.Join(n, ", \n\t")
+	tableName := table
+	createTableSQL := `CREATE TABLE ` + tableName + ` (` + "\n\t" + fields + "\n" + `);`
+
+	t.mustExecute(createTableSQL)
+	fmt.Printf("Table %s created successfully\n", tableName)
+
+	storedTasks := []any{sampleEntry}
+	t.tables[tableName] = storedTasks
+	t.entryCount++
+	if t.entryCount >= t.batchSize {
+		t.Flush()
+	}
+}
+
+func (t *SQLiteWriter) DataInsert(table string, entry any) {
+	err := t.checkStructFields(entry)
+	if err != nil {
+		panic(err)
+	}
+
+	storedTasks, exists := t.tables[table]
+	if !exists {
+		panic(fmt.Errorf("table %s does not exist", table))
+	}
+
+	stdTask := storedTasks[0]
```

Comment:
Can you guarantee that there is always an entry in the `storedTask` field? Or do you just reserve spot 0 for stdTask? I would consider having a dedicated field that maps from the table name to the stdTask.

Reply:
Yes, spot 0 is reserved for stdTask.